File: src/api/middleware/authMiddleware.js
/**
 * Authentication Middleware
 * Validates JWT tokens and attaches user to request
 */

const logger = require('../../../logger');

/**
 * Create authentication middleware
 * @param {ValidateToken} validateTokenUseCase - ValidateToken use case
 * @returns {Function} Express middleware
 */
function createAuthMiddleware(validateTokenUseCase) {
    return async (req, res, next) => {
        try {
            // Extract token from Authorization header
            const authHeader = req.headers.authorization;

            if (!authHeader) {
                return res.status(401).json({
                    error: 'Missing authorization header',
                    code: 'MISSING_AUTH_HEADER'
                });
            }

            // Expect format: "Bearer <token>"
            const parts = authHeader.split(' ');
            if (parts.length !== 2 || parts[0] !== 'Bearer') {
                return res.status(401).json({
                    error: 'Invalid authorization header format',
                    code: 'INVALID_AUTH_FORMAT',
                    details: {
                        expected: 'Bearer <token>'
                    }
                });
            }

            const token = parts[1];

            // Validate token
            const result = await validateTokenUseCase.execute({ token });

            // Attach user to request
            req.user = result.user;

            next();
        } catch (error) {
            logger.warn('Authentication failed', {
                error: error.message,
                path: req.path,
                ip: req.ip
            });

            return res.status(401).json({
                error: 'Invalid or expired token',
                code: 'INVALID_TOKEN'
            });
        }
    };
}

/**
 * Create optional authentication middleware
 * Attaches user if token is valid, but doesn't require it
 * @param {ValidateToken} validateTokenUseCase - ValidateToken use case
 * @returns {Function} Express middleware
 */
function createOptionalAuthMiddleware(validateTokenUseCase) {
    return async (req, res, next) => {
        try {
            const authHeader = req.headers.authorization;

            if (!authHeader) {
                // No token provided, continue without user
                return next();
            }

            const parts = authHeader.split(' ');
            if (parts.length === 2 && parts[0] === 'Bearer') {
                const token = parts[1];

                try {
                    const result = await validateTokenUseCase.execute({ token });
                    req.user = result.user;
                } catch (error) {
                    // Token validation failed, but that's okay for optional auth
                    logger.debug('Optional auth failed', { error: error.message });
                }
            }

            next();
        } catch (error) {
            // Don't fail the request, just continue without user
            logger.error('Optional auth middleware error', {
                error: error.message,
                path: req.path
            });
            next();
        }
    };
}

module.exports = {
    createAuthMiddleware,
    createOptionalAuthMiddleware
};

File: zapzap-rust/src/api/middleware/auth_middleware.rs
use std::sync::Arc;

use axum::{
    extract::{Request, State},
    http::{header::AUTHORIZATION, HeaderMap, StatusCode},
    middleware::Next,
    response::{IntoResponse, Json, Response},
};

use crate::domain::repositories::UserRepository;
use crate::infrastructure::app_state::AppState;

// Re-export Claims for use in route handlers
pub use crate::infrastructure::auth::Claims;

/// Whether the token's user is still in the database; a lookup error counts as no.
pub async fn user_exists(state: &AppState, user_id: &str) -> bool {
    match state.user_repo.find_by_id(user_id).await {
        Ok(user) => user.is_some(),
        Err(e) => {
            tracing::error!("Auth user lookup failed: {}", e);
            false
        }
    }
}

/// Why a request carries no usable bearer token.
enum BearerError {
    /// No `Authorization` header, or an empty one (Express reads it as absent)
    Missing,
    /// A header that is not exactly `Bearer <token>`
    Malformed,
}

/// The token of an `Authorization: Bearer <token>` header, read as the Node backend's
/// middlewares did: split on single spaces, exactly two parts, the first `Bearer`.
fn bearer_token(headers: &HeaderMap) -> Result<&str, BearerError> {
    let header = match headers.get(AUTHORIZATION) {
        Some(h) if !h.is_empty() => h,
        _ => return Err(BearerError::Missing),
    };
    let text = header.to_str().map_err(|_| BearerError::Malformed)?;
    match text.split(' ').collect::<Vec<_>>().as_slice() {
        ["Bearer", token] => Ok(token),
        _ => Err(BearerError::Malformed),
    }
}

/// A 401 in the Node backend's shape: `{error, code, details?}`.
fn unauthorized(error: &str, code: &str, details: Option<serde_json::Value>) -> Response {
    let mut body = serde_json::json!({ "error": error, "code": code });
    if let Some(details) = details {
        body["details"] = details;
    }
    (StatusCode::UNAUTHORIZED, Json(body)).into_response()
}

/// Extract authenticated user from request. Refusals match the Node backend's:
/// no header is `MISSING_AUTH_HEADER`, a header that is not exactly `Bearer <token>` is
/// `INVALID_AUTH_FORMAT`, and a bad or expired token, or one whose user no longer exists
/// (the Node backend answered the same), is `INVALID_TOKEN`.
pub async fn auth_middleware(
    State(state): State<Arc<AppState>>,
    mut request: Request,
    next: Next,
) -> Response {
    let invalid_token = || unauthorized("Invalid or expired token", "INVALID_TOKEN", None);

    let claims = match bearer_token(request.headers()) {
        Ok(token) => match state.jwt_service.verify(token) {
            Ok(claims) => claims,
            Err(_) => return invalid_token(),
        },
        Err(BearerError::Missing) => {
            return unauthorized("Missing authorization header", "MISSING_AUTH_HEADER", None)
        }
        Err(BearerError::Malformed) => {
            return unauthorized(
                "Invalid authorization header format",
                "INVALID_AUTH_FORMAT",
                Some(serde_json::json!({ "expected": "Bearer <token>" })),
            )
        }
    };

    // The user must still exist, one primary-key lookup
    if !user_exists(&state, &claims.user_id).await {
        return invalid_token();
    }

    // Add claims to request extensions
    request.extensions_mut().insert(claims);

    next.run(request).await
}

/// Optional auth middleware - doesn't fail if no token: a well-formed header with a valid
/// token of an existing user attaches the claims, anything else goes on without them
pub async fn optional_auth_middleware(
    State(state): State<Arc<AppState>>,
    mut request: Request,
    next: Next,
) -> Response {
    let claims = bearer_token(request.headers())
        .ok()
        .and_then(|token| state.jwt_service.verify(token).ok());
    if let Some(claims) = claims {
        if user_exists(&state, &claims.user_id).await {
            request.extensions_mut().insert(claims);
        }
    }

    next.run(request).await
}

/// Admin middleware - requires an authenticated user who is an admin *now*: the flag is
/// read from the database, not from the token, so a revoked admin loses access at once
/// (the Node backend's statuses, messages and codes). Mounted inside
/// `auth_middleware`, which sets the claims.
pub async fn admin_middleware(
    State(state): State<Arc<AppState>>,
    request: Request,
    next: Next,
) -> Response {
    let refuse = |status: StatusCode, error: &str, code: &str| {
        let body = serde_json::json!({ "success": false, "error": error, "code": code });
        (status, Json(body)).into_response()
    };

    let Some(claims) = request.extensions().get::<Claims>() else {
        return refuse(
            StatusCode::UNAUTHORIZED,
            "Authentication required",
            "AUTH_REQUIRED",
        );
    };

    match state.user_repo.find_by_id(&claims.user_id).await {
        Ok(Some(user)) if user.is_admin => next.run(request).await,
        Ok(_) => refuse(
            StatusCode::FORBIDDEN,
            "Admin access required",
            "ADMIN_REQUIRED",
        ),
        Err(e) => {
            tracing::error!("Admin middleware error: {}", e);
            refuse(
                StatusCode::INTERNAL_SERVER_ERROR,
                "Internal server error",
                "INTERNAL_ERROR",
            )
        }
    }
}
